cloud/ec2.go

```diff
@@ -1167,66 +1167,57 @@ func (m *ec2Manager) TerminateInstance(ctx context.Context, h *host.Host, user,
 
 // StopInstance stops a running EC2 instance.
 func (m *ec2Manager) StopInstance(ctx context.Context, h *host.Host, user string) error {
-	// Check if already stopped or not running
 	if h.Status == evergreen.HostStopped {
 		return errors.Errorf("cannot stop '%s' - already marked as stopped", h.Id)
-	} else if h.Status != evergreen.HostRunning {
-		return errors.Errorf("cannot stop '%s' - host is not running", h.Id)
+	} else if h.Status != evergreen.HostRunning && h.Status != evergreen.HostStopping {
+		return errors.Errorf("cannot stop '%s' - host status is '%s'", h.Id, h.Status)
 	}
 
 	if err := m.client.Create(m.credentials, m.region); err != nil {
-		return errors.Wrap(err, "error creating client")
+		return errors.Wrap(err, "creating client")
 	}
 	defer m.client.Close()
 
-	var instance *ec2.Instance
-	instance, err := m.client.GetInstanceInfo(ctx, h.Id)
-	if err != nil {
-		return errors.Wrap(err, "error getting instance info before stopping")
-	}
-
-	// if instance is already stopped, return early
-	if ec2StatusToEvergreenStatus(*instance.State.Name) == StatusStopped {
-		grip.Info(message.Fields{
-			"message":       "instance already stopped",
-			"user":          user,
-			"host_provider": h.Distro.Provider,
-			"host_id":       h.Id,
-			"distro":        h.Distro.Id,
-		})
-
-		return errors.Wrap(h.SetStopped(user), "failed to mark instance as stopped in db")
-	}
-
-	if err = h.SetStopping(user); err != nil {
-		return errors.Wrap(err, "failed to mark instance as stopping in db")
-	}
-
-	_, err = m.client.StopInstances(ctx, &ec2.StopInstancesInput{
+	out, err := m.client.StopInstances(ctx, &ec2.StopInstancesInput{
 		InstanceIds: []*string{aws.String(h.Id)},
 	})
 	if err != nil {
-		// fetch the instance to get the current status, since it may have changed
-		// since we saved the previous status
-		currentStatus, currentStatusErr := m.GetInstanceStatus(ctx, h)
-		if currentStatusErr != nil {
-			return errors.Wrap(currentStatusErr, "error getting instance status after stopping error")
-		}
-
-		if currentStatusErr = h.SetStatusAtomically(currentStatus.String(), user, "reverting to the most up-to-date host status"); currentStatusErr != nil {
-			return errors.Wrapf(currentStatusErr, "failed to revert status from stopping to '%s'", currentStatus)
-		}
-		return errors.Wrapf(err, "error stopping EC2 instance '%s'", h.Id)
+		return errors.Wrapf(err, "stopping EC2 instance '%s'", h.Id)
 	}
 
-	// Check whether instance stopped
+	if len(out.StoppingInstances) == 1 {
+		instance := out.StoppingInstances[0]
+		status := ec2StatusToEvergreenStatus(utility.FromStringPtr(instance.CurrentState.Name))
+		switch status {
+		case StatusStopping:
+			grip.Error(message.WrapError(h.SetStopping(user), message.Fields{
+				"message": "could not mark host as stopping, continuing to poll instance status anyways",
```

Comment:
Probably. But if that happens, two things can happen to the job, both of which seem acceptable:
* The job also fails to update the Amboy DB, in which case it'll re-dispatch after the stale lock timeout and try doing this again.
* The job successfully updates the Amboy DB and logs an event to the host events saying the failure reason, at which point the user could see the error message in the host event page and press the stop button again if they wanted to try again.

Reply:
but I mean to say: you mentioned it was an explicit decision not to return here. I was wondering why not return now if we can't set a state on the host.

Reply:
Mostly because "stopping" is only a nice to have state for the user's benefit and is not necessary for `StopInstances` to be considered a success. It's still possible that there's something that prevents any DB writes, but if it's just a momentary blip (or something else that prevents `SetStopping` from updating), I don't want it to give up here just because the user doesn't see the "stopping" state. Once it's reached this point, the cloud host is really stopping, so it's more useful to at least have debug logs or more information by continuing on error.